### PR TITLE
Fix bug that breaks PPO's value bootstrapping when a terminal reward is present.

### PR DESCRIPTION
**Pull request by MatthewCWeston**
## Description
At present, `compute_value_targets` zeros out the values of states that are followed by a termination before calculating the intermediates used for value bootstrapping. This means that, under pure value bootstrapping _(`lambda_=0`)_, state A, which allows an agent to access a large reward immediately and terminate, will have an appropriate value target, but state B, which allows an agent to access state A, will have a value target of zero. In effect, this causes configs with `lambda_<1` to perform worse than expected in any environment with terminal rewards, and breaks learning entirely when `lambda_=0`.

This PR modifies `AddOneTsToEpisodesAndTruncate` to fix the issue, and implements a lightweight test to prevent regressions.

## Related issues
Resolves the bug outlined in #57683.

### CODE REVIEW

File: rllib/algorithms/ppo/tests/test_ppo_value_bootstrapping.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+import ray
+import ray.rllib.algorithms.ppo as ppo
+from ray.rllib.connectors.env_to_module import FlattenObservations
+from ray.rllib.core.columns import Columns
+from ray.rllib.policy.sample_batch import DEFAULT_POLICY_ID
+from ray.rllib.utils.metrics import ENV_RUNNER_RESULTS, EPISODE_RETURN_MEAN
+import torch
+
+class TestPPO(unittest.TestCase):
```
[the remark this answers is not in the document]
Comment (MatthewCWeston) [reply rllib/algorithms/ppo/tests/test_ppo_value_bootstrapping.py:12]: The current naming scheme is consistent with the other PPO test scripts in the same folder. Happy to make the change if a maintainer requests it.

File: rllib/algorithms/ppo/tests/test_ppo_value_bootstrapping.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        )
+
+        num_iterations = 20
+
+        algo = config.build()
+
+        for i in range(num_iterations):
+            r_mean = algo.train()[ENV_RUNNER_RESULTS][EPISODE_RETURN_MEAN]
+            print(r_mean)
+
```
[the remark this answers is not in the document]
Comment (MatthewCWeston) [reply rllib/algorithms/ppo/tests/test_ppo_value_bootstrapping.py]: Test scripts' outputs only display if the test fails. I've tried to match the verbosity of the other tests and include information that would be essential in quickly determining the root cause of a test case failure.

File: rllib/algorithms/ppo/tests/test_ppo_value_bootstrapping.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        critic = algo.learner_group._learner._module[DEFAULT_POLICY_ID]
+        state_values = {}
+ 
+        for state in [3,2,4,6]:
+            obs = torch.zeros((8,)).float()
+            obs[state]+=1
+            batch = {Columns.OBS: obs.unsqueeze(0)}
+            with torch.no_grad():
+              value = critic.compute_values(batch).item()
+            print(f'State {state}: {value:.02f}')
```
[the remark this answers is not in the document]
Comment (MatthewCWeston) [reply rllib/algorithms/ppo/tests/test_ppo_value_bootstrapping.py]: Indentation: Forgot to add changes after linting. Will fix immediately.

Print statement: Test scripts' outputs only display if the test fails. I've tried to match the verbosity of the other tests and include information that would be essential in quickly determining the root cause of a test case failure.

File: rllib/algorithms/ppo/tests/test_ppo_value_bootstrapping.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+            state_values[state] = value
+
+        algo.stop()
+        # Value bootstrapping should learn this simple environment reliably
+        self.assertGreater(r_mean, 0.9)
+        # The value function
+        self.assertGreater(state_values[3], 0.9) # Immediately terminates with reward 1
+        self.assertGreater(state_values[2], 0.8) # One step from terminating with reward 1
+        self.assertGreater(state_values[4], 0.7) # Two steps from terminating with reward 1
+        self.assertLess(state_values[6], 0.7) # Cannot reach the goal from this state
```
[the remark this answers is not in the document]
Comment (MatthewCWeston) [reply rllib/algorithms/ppo/tests/test_ppo_value_bootstrapping.py]: This seems to actively make the targets less clear, though happy to adjust wording if a human reviewer agrees that it's presently too vague. The optimal values of the states, with gamma=0.9 and θ*, is:

```
3: 1.00 (In this state, can immediately terminate with reward 1.)
2: 0.90 (In this state, can transition to state 3.)
4: 0.81 (In this state, can transition to state 2, then to 3.)
6: 0.00 (In this state, cannot obtain positive reward.)
```

_(The asserts leave room for as much noise as they can while still verifying that the value function is reasonable)_

File: rllib/algorithms/ppo/tests/test_ppo_value_bootstrapping.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (MatthewCWeston) [reply rllib/algorithms/ppo/tests/test_ppo_value_bootstrapping.py]: Fixed.

**Comment by pseudo-rnd-thoughts** [comment]
Thanks for the PR, looking at the test, I don't think that PPO is the problem. 
I believe the problem is actually that `max_episode_steps=3` which will cause most episode to truncate almost everytime.
Updating this value to `50` and training for longer then I can get a `r_mean` of `0.99`. 
Further, printing all the states, I get
```
State 0: 0.11
State 1: 0.37
State 2: 0.00
State 3: 1.00
State 4: -0.00
State 5: 0.08
State 6: 0.00
State 7: 0.21
```

@MatthewCWeston Could you clarify what you think the core bug is and what you fix is

**Comment by MatthewCWeston** [comment]
@pseudo-rnd-thoughts The bug, as I understand it, is that the value target prediction cannot handle terminal rewards when value bootstrapping - the value of the last state observed in an episode doesn't get used to inform the value target of the state before it. I fix the bug in this PR _(by modifying `add_one_ts_to_episodes_and_truncate.py` to move the termination flag to the artificial timestep that it adds; I believe this not being done in the first place was accidental)_, but reverting this change will cause the new test to fail.

I've written a deeper explanation of the bug itself on the related issue (https://github.com/ray-project/ray/issues/57683), with a replication script that should be directly runnable on the base repo to verify the behavior in question [here](https://github.com/user-attachments/files/23001683/bug_lambda_demo.py). 

With the change I've made, the test should already pass reliably without the need to adjust the time horizon. Does the test pass without my fix if the time horizon is increased? It didn't when I tested it with multiple seeds; I've looked at the value targets directly, and they are 1.0 for the last observed state and roughly 0.0 for all other states when my modification isn't present. The value function shouldn't learn at all under those conditions, and the test checks for _(approximate)_ value function correctness. 

It is possible, though not guaranteed, to get a high mean reward without this, if the policy network's initial parameters are lucky, since we're not actively making them worse and we are learning the correct action in state 2 because the reward is immediate, but the state values you provided, under an optimal policy with `γ=0.9`, should look like this:

```
2: 0.90 (0.0 is definitely incorrect; we're one timestep away from a reward of 1.0)
3: 1.00 (1.0 is the correct value, but it doesn't get used to set state 2's value target)
4: 0.81 (likewise, the value of the starting position shouldn't be zero, since we can reach the goal in two steps)
6: 0.0 (correct; even a different value isn't cause for alarm, since this state doesn't appear at all under θ*)

(2, 3, 4, and 6 are the only reachable states, so the other states can have arbitrary value)
```

**Comment by pseudo-rnd-thoughts** [comment]
Thanks @MatthewCWeston, I agree that a bug most likely exists, however I don't think that the proposed change in `rllib/connectors/learner/add_one_ts_to_episodes_and_truncate.py` is correct as it treats all episodes as terminated. 

This is a problem for PPO as when it does a rollout of a fixed length, it can collect partial episodes. In this case then the proposed change is causing the agent to believe the partial episode actually did terminate for a "random" state. 

The test though is very helpful in understanding that the agent isn't learning as expected and that the bug is most likely in how the agent learns on these partial episodes. 
I don't have time currently to look at it for the next two weeks but I can look at it after then

**Comment by MatthewCWeston** [comment]
@pseudo-rnd-thoughts 

**Edit:**

I had a look at [value_predictions.py](https://github.com/ray-project/ray/blob/fe7ad00f9720a722fde5fecba5bb681234bcdb63/rllib/utils/postprocessing/value_predictions.py), and I think the edge case you describe is already handled. I don't alter the truncated flags at all, and the value target prediction code checks for truncated episodes and overrides the value targets at the last step with the current value predictions appropriately, which is more efficient than the change I described in the original version of this comment.


~I think you're right about that being an edge-case, but it's present in the existing code. The terminated flag in the new timestep isn't something I added; I simply removed the (potential) duplicate terminated flag in the original final step.~

~It could, I believe, be resolved by masking out (via `Columns.LOSS_MASK`) the last _(real)_ timestep in the case that an episode is truncated rather than terminated. To accurately calculate its value target, we'd need to know the value of the subsequent state, which we don't have yet, so that'd be all we can do in that case.~

**Comment by MatthewCWeston** [comment]
I think this is scheduled to get another look at some point, and as far as I can tell, the bug persists, so I'll comment here to prevent it from being marked as stale.

**Review by HassamSheikh — COMMENTED** [review]
I agree with @pseudo-rnd-thoughts  comment here, this change will consider any random state especially in long horizon problem as a terminal goal/failure state.

**Comment by MatthewCWeston** [comment]
@HassamSheikh I think I addressed that above:

> I had a look at [value_predictions.py](https://github.com/ray-project/ray/blob/fe7ad00f9720a722fde5fecba5bb681234bcdb63/rllib/utils/postprocessing/value_predictions.py), and I think the edge case you describe is already handled. I don't alter the truncated flags at all, and the value target prediction code checks for truncated episodes and overrides the value targets at the last step with the current value predictions appropriately, which is more efficient than the change I described in the original version of this comment

As it is, the in-production code already adds the terminated flag to partial episodes, but that does not mean that GAE treats them as if they were terminated. `value_predictions.py` uses the truncated flags to apply VF estimates to the ends of partial episodes, and my code doesn't alter the presence or absence of those flags. 

Put another way, the only thing adjusted by this PR should by the behavior of non-truncated episodes with terminal rewards. The behavior of truncated episodes, as I understand it, is completely unchanged. If I'm wrong about this, and there's a case where this isn't true, I'd be happy to patch that out.

**Comment by pseudo-rnd-thoughts** [comment]
Reviewing the code for the GAE implementation, then the artificial time step that added by the `AddOneTsToEpisodesAndTruncate` has `terminated=True` as we use it to initialise the GAE value at 0, then depending on if it is a partial / truncated episode or a complete terminated episode the second to last needs to be conditional which this PR changes to always be True. 

Therefore, I think this implementation does affect truncated episodes.  

I'm closing the PR but happy to continue the conversation

**Comment by MatthewCWeston** [comment]
@pseudo-rnd-thoughts I went in to check manually, and the behavior you describe does indeed occur. However, it also occurred when I ran the same test without the PR  - it appears to be due to the new last timestep always having the terminated flag, rather than my change to the original last timestep's terminated flag. Code is below, in case I've made an incorrect assumption somewhere:

```
from ray.rllib.utils.postprocessing.value_predictions import compute_value_targets
from ray.rllib.utils.postprocessing.zero_padding import (
    split_and_zero_pad_n_episodes,
    unpad_data_if_necessary,
)

import numpy as np

# Compute value targets for GAE
def cvt(episode_lens, vfps, rewards, terminateds, truncateds, gamma=0.99, lambda_=0.0):
  return compute_value_targets(
      values=vfps,
      rewards=unpad_data_if_necessary(
          episode_lens,
          np.array(rewards),
      ),
      terminateds=unpad_data_if_necessary(
          episode_lens,
          np.array(terminateds),
      ),
      truncateds=unpad_data_if_necessary(
          episode_lens,
          np.array(truncateds),
      ),
      gamma=gamma,
      lambda_=lambda_,
  )

print('''
===== CURRENT PRODUCTION CODE =====
''')
# First episode is terminated, second episode is terminated, current production code.
# Correct VTs: [0.9405, 1.,     <any>,     0.9405, 1.,     <any>    ]
# Actual VTs:  [0. 1. 0. 0. 1. 0.] # Final reward missed by earlier states!
print("First episode is terminated, second episode is terminated:")
print(cvt(
    [3,3], # Episode lengths
    [0.0, 0.95, 0.95, 0.0, 0.95, 0.95], # Value head outputs
    [ 0.0, 1.0,  0.0,  0.0,  1.0,  0.0], # Environment rewards
    # In production code, the dummy last timestep always has the terminated flag, and the real one has it if the episode terminated.
    [False,  True,  True, False,  True,  True], # Terminated flags
    [False, False, False, False, False, False], # Truncated flags
))
# First episode is terminated, second episode is truncated, current production code.
# Correct VTs: [0.9405, 1.,     <any>,     0.9405, 1.9405,     <any>    ]
# Actual VTs:  [0.     1.     0.     0.9405 1.     0.    ]
print("First episode is terminated, second episode is truncated:")
print(cvt(
    [3,3], # Episode lengths
    [0.0, 0.95, 0.95, 0.0, 0.95, 0.95], # Value head outputs
    [ 0.0, 1.0,  0.0,  0.0,  1.0,  0.0], # Environment rewards
    # In production code, the dummy last timestep always has the terminated flag, and the real one has it if the episode terminated.
    [False,  True,  True, False,  False,  True], # Terminated flags
    [False, False, False, False, False, True], # Truncated flags
))

print('''
========== WITH THIS PR ===========
''')
# First episode is terminated, second episode is terminated, PR
# Correct VTs: [0.9405, 1.,     <any>,     0.9405, 1.,     <any>    ]
# Actual VTs:  [0.9405 1.     0.     0.9405 1.     0.    ]
print("First episode is terminated, second episode is terminated:")
print(cvt(
    [3,3], # Episode lengths
    [0.0, 0.95, 0.95, 0.0, 0.95, 0.95], # Value head outputs
    [ 0.0, 1.0,  0.0,  0.0,  1.0,  0.0], # Environment rewards
    # In the PR, the terminated flag is removed from the true last timestep no matter what.
    [False,  False,  True, False,  False,  True], # Terminated flags
    [False, False, False, False, False, False], # Truncated flags
))
# First episode is terminated, second episode is truncated, PR
# Correct VTs: [0.9405, 1.,     <any>,     0.9405, 1.9405,     <any>    ]
# Actual VTs:  [0.9405 1.     0.     0.9405 1.     0.    ]
print("First episode is terminated, second episode is truncated:")
print(cvt(
    [3,3], # Episode lengths
    [0.0, 0.95, 0.95, 0.0, 0.95, 0.95], # Value head outputs
    [ 0.0, 1.0,  0.0,  0.0,  1.0,  0.0], # Environment rewards
    # In the PR, the terminated flag is removed from the true last timestep no matter what.
    [False,  False,  True, False,  False,  True], # Terminated flags
    [False, False, False, False, False, True], # Truncated flags
))
print('''
========= PROPOSED CHANGE =========
''')
# First episode is terminated, second episode is terminated, proposed change
# Correct VTs: [0.9405, 1.,     <any>,     0.9405, 1.,     <any>    ]
# Actual VTs:  [0.9405 1.     0.     0.9405 1.     0.    ]
print("First episode is terminated, second episode is terminated:")
print(cvt(
    [3,3], # Episode lengths
    [0.0, 0.95, 0.95, 0.0, 0.95, 0.95], # Value head outputs
    [ 0.0, 1.0,  0.0,  0.0,  1.0,  0.0], # Environment rewards
    # Proposal: The new last timestep has the terminated flag only if the original last timestep had it.
    [False,  False,  True, False,  False,  True], # Terminated flags
    [False, False, False, False, False, False], # Truncated flags
))
# First episode is terminated, second episode is truncated, proposed change
# Correct VTs: [0.9405, 1.,     <any>,     0.9405, 1.9405,     <any>    ]
# Actual VTs:  [0.9405 1.     0.     0.9405 1.9405 0.    ]
print("First episode is terminated, second episode is truncated:")
print(cvt(
    [3,3], # Episode lengths
    [0.0, 0.95, 0.95, 0.0, 0.95, 0.95], # Value head outputs
    [ 0.0, 1.0,  0.0,  0.0,  1.0,  0.0], # Environment rewards
    # Proposal: The new last timestep has the terminated flag only if the original last timestep had it.
    [False,  False,  True, False,  False,  False], # Terminated flags
    [False, False, False, False, False, True], # Truncated flags
))
```

It seems like what we want is to set the _new_ last timestep's terminated flag to `False` if the episode is truncated, rather than always setting it to `True`. Does that sound right to you? If so, I can modify the PR accordingly.

**Comment by pseudo-rnd-thoughts** [comment]
> It seems like what we want is to set the new last timestep's terminated flag to False if the episode is truncated, rather than always setting it to True. Does that sound right to you? If so, I can modify the PR accordingly.

I've a tad confused as thats what I believe the current implementation does
```python
terminateds = (
   [False for _ in range(len_ - 1)]
   + [bool(sa_episode.is_terminated)]
   + [True]  # extra timestep
)
```

**Comment by MatthewCWeston** [comment]
@pseudo-rnd-thoughts Wording might've been unclear - 'last timestep' refers to the artificial one. In terms of actual code, I think it'd look like:

```
terminateds = (
                [False for _ in range(len_ - 1)]
                + [False] # FIX: Setting to True if terminated causes last-step rewards to be ignored!
                + [bool(sa_episode.is_terminated)] # FIX: Setting to True appears to break VT calculation for truncated episodes
            )
```

The tests above show what value target calculation looks like under each of the three paradigms (production code, original PR, and the rules above). This is the one that seems to get us the correct targets.
